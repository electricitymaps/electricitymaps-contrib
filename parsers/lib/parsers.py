import importlib

from electricitymap.contrib.config import EXCHANGES_CONFIG, ZONES_CONFIG

# Prepare all parsers
CONSUMPTION_PARSERS = {}
PRODUCTION_PARSERS = {}
PRODUCTION_PER_MODE_FORECAST_PARSERS = {}
PRODUCTION_PER_UNIT_PARSERS = {}
EXCHANGE_PARSERS = {}
PRICE_PARSERS = {}
CONSUMPTION_FORECAST_PARSERS = {}
GENERATION_FORECAST_PARSERS = {}
EXCHANGE_FORECAST_PARSERS = {}
PRODUCTION_CAPACITY_PARSERS = {}
DAYAHEAD_LOCATIONAL_MARGINAL_PRICE_PARSERS = {}
REALTIME_LOCATIONAL_MARGINAL_PRICE_PARSERS = {}

PARSER_KEY_TO_DICT = {
    "consumption": CONSUMPTION_PARSERS,
    "production": PRODUCTION_PARSERS,
    "productionPerUnit": PRODUCTION_PER_UNIT_PARSERS,
    "productionPerModeForecast": PRODUCTION_PER_MODE_FORECAST_PARSERS,
    "exchange": EXCHANGE_PARSERS,
    "price": PRICE_PARSERS,
    "dayaheadLocationalMarginalPrice": DAYAHEAD_LOCATIONAL_MARGINAL_PRICE_PARSERS,
    "realtimeLocationalMarginalPrice": REALTIME_LOCATIONAL_MARGINAL_PRICE_PARSERS,
    "consumptionForecast": CONSUMPTION_FORECAST_PARSERS,
    "generationForecast": GENERATION_FORECAST_PARSERS,
    "exchangeForecast": EXCHANGE_FORECAST_PARSERS,
    "productionCapacity": PRODUCTION_CAPACITY_PARSERS,
}


def _parser_key_to_parser_folder(parser_key: str):
    return (
        "electricitymap.contrib.capacity_parsers"
        if parser_key == "productionCapacity"
        else "parsers"
    )


# Read all zones
for zone_id, zone_config in ZONES_CONFIG.items():
    for parser_key, v in zone_config.get("parsers", {}).items():
        mod_name, fun_name = v.split(".")
        mod = importlib.import_module(
            f"{_parser_key_to_parser_folder(parser_key)}.{mod_name}"
        )
        PARSER_KEY_TO_DICT[parser_key][zone_id] = getattr(mod, fun_name)


# Read all exchanges
for exchange_id, exchange_config in EXCHANGES_CONFIG.items():
    for parser_key, v in exchange_config.get("parsers", {}).items():
        mod_name, fun_name = v.split(".")
        mod = importlib.import_module(f"parsers.{mod_name}")
        PARSER_KEY_TO_DICT[parser_key][exchange_id] = getattr(mod, fun_name)
