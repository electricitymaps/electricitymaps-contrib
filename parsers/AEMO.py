import io
import re
import zipfile
from datetime import datetime
from logging import Logger, getLogger
from typing import Any
from urllib.parse import urljoin
from zoneinfo import ZoneInfo

import pandas as pd
from bs4 import BeautifulSoup
from requests import Session

from electricitymap.contrib.lib.models.event_lists import TotalConsumptionList
from electricitymap.contrib.lib.models.events import EventSourceType
from electricitymap.contrib.lib.types import ZoneKey

SOURCE = "aemo.com.au"

ZONE_KEY_TO_REGION = {
    "AU-NSW": "NSW1",
    "AU-QLD": "QLD1",
    "AU-SA": "SA1",
    "AU-TAS": "TAS1",
    "AU-VIC": "VIC1",
    "AU-WA": "WEM",  # This zone is not implemented yet
}

ZONE_KEY_TO_TIMEZONE = {
    "AU-NSW": ZoneInfo("Australia/Sydney"),
    "AU-QLD": ZoneInfo("Australia/Brisbane"),
    "AU-SA": ZoneInfo("Australia/Adelaide"),
    "AU-TAS": ZoneInfo("Australia/Hobart"),
    "AU-VIC": ZoneInfo("Australia/Melbourne"),
    "AU-WA": ZoneInfo("Australia/Perth"),  # This zone is not implemented yet
}

# TODO, what about the other zone in Australia (AU-WA)? Check remaining zone


def find_document(session, target_datetime):
    # Fetch the directory listing
    base_url = "http://nemweb.com.au/Reports/CURRENT/Operational_Demand/FORECAST_HH/"
    response = session.get(base_url)

    # Parse with BeautifulSoup
    soup = BeautifulSoup(response.text, "html.parser")

    # String datetime
    target_date_str = target_datetime.strftime("%Y%m%d")

    # Find matching links
    matching_links = soup.find_all(
        "a",
        href=re.compile(
            rf"PUBLIC_FORECAST_OPERATIONAL_DEMAND_HH_{target_date_str}\d+_\d+\.zip"
        ),
    )

    if matching_links:
        target_link = matching_links[-1]  # Get the last (most recent) link
        full_url = urljoin(base_url, target_link["href"])  # Construct full URL

        file_response = session.get(full_url)

        with zipfile.ZipFile(io.BytesIO(file_response.content)) as z:
            csv_filename = z.namelist()[0]
            with z.open(csv_filename) as f:
                df = pd.read_csv(f)
                return df
    else:
        print("No matching files found")


def fetch_consumption_forecast(
    zone_key: ZoneKey,  # "AU-NSW", "AU-QLD", "AU-SA", "AU-TAS", "AU-VIC", "AU-WA"
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict[str, Any]]:
    """Consumption forecast in MW every half an hour for 10 days ahead.
    Only for NSW1, QND1, SA1, TAS1, VIC1 zones."""
    session = session or Session()

    if target_datetime is None:
        target_datetime = datetime.now(tz=ZONE_KEY_TO_TIMEZONE[zone_key])

    df = find_document(session, target_datetime)

    # Transform dataframe
    df.columns = df.iloc[0]
    df = df.iloc[1:-1].reset_index(drop=True)

    #
    region = ZONE_KEY_TO_REGION.get(zone_key)
    all_consumption_events = df[
        df["REGIONID"] == region
    ]  # all events with a datetime and a consumption value
    consumption_list = TotalConsumptionList(logger)
    for _, event in all_consumption_events.iterrows():
        datetime_object = datetime.strptime(
            event["INTERVAL_DATETIME"], "%Y/%m/%d %H:%M:%S"
        ).replace(tzinfo=ZONE_KEY_TO_TIMEZONE[zone_key])

        consumption_list.append(
            zoneKey=zone_key,
            datetime=datetime_object,
            consumption=float(
                event["OPERATIONAL_DEMAND_POE50"]
            ),  # 50% probability of exceedance operational demand forecast value
            source=SOURCE,
            sourceType=EventSourceType.forecasted,
        )
    return consumption_list.to_list()


if __name__ == "__main__":
    """Main method, never used by the electricityMap backend, but handy for testing."""

    print(fetch_consumption_forecast("AU-NSW"))
    print(fetch_consumption_forecast("AU-QLD"))
    print(fetch_consumption_forecast("AU-SA"))
    print(fetch_consumption_forecast("AU-TAS"))
    print(fetch_consumption_forecast("AU-VIC"))
    print(
        fetch_consumption_forecast("AU-WA")
    )  # Not implemented yet. It returns an empty list
