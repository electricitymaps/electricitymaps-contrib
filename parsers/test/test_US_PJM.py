import re
from json import loads
from pathlib import Path

from requests_mock import GET

from electricitymap.contrib.lib.types import ZoneKey
from parsers.US_PJM import fetch_production, fetch_wind_solar_forecasts

base_path_to_mock = Path("parsers/test/mocks/US_PJM")


def test_production(adapter, session, snapshot):
    settings = Path(base_path_to_mock, "settings.json")
    adapter.register_uri(
        GET,
        "https://dataminer2.pjm.com/config/settings.json",
        json=loads(settings.read_text()),
    )

    data = Path(base_path_to_mock, "gen_by_fuel.json")
    adapter.register_uri(
        GET,
        re.compile(
            r"https://us-ca-proxy-jfnx5klx2a-uw\.a\.run\.app/api/v1/gen_by_fuel.*"
        ),
        json=loads(data.read_text()),
    )

    assert snapshot == fetch_production(
        zone_key=ZoneKey("US-MIDA-PJM"),
        session=session,
    )


def test_fetch_wind_solar_forecasts(adapter, session, snapshot):
    # Mock the settings.json request
    settings = Path(base_path_to_mock, "settings.json")
    adapter.register_uri(
        GET,
        "https://dataminer2.pjm.com/config/settings.json",
        json=loads(settings.read_text()),
    )

    # Mock wind forecast request
    data_wind = Path(
        base_path_to_mock, "pjm_wind_forecast_2025-02-24T00:00:00.0000000.json"
    )
    adapter.register_uri(
        GET,
        re.compile(
            r"https://us-ca-proxy-jfnx5klx2a-uw\.a\.run\.app/api/v1/hourly_wind_power_forecast.*"
        ),
        json=loads(data_wind.read_text()),
    )

    # Mock solar forecast request
    data_solar = Path(
        base_path_to_mock, "pjm_solar_forecast_2025-02-24T00:00:00.0000000.json"
    )
    adapter.register_uri(
        GET,
        re.compile(
            r"https://us-ca-proxy-jfnx5klx2a-uw\.a\.run\.app/api/v1/hourly_solar_power_forecast.*"
        ),
        json=loads(data_solar.read_text()),
    )

    # Run function under test
    assert snapshot == fetch_wind_solar_forecasts(
        zone_key=ZoneKey("US-MIDA-PJM"),
        session=session,
    )
