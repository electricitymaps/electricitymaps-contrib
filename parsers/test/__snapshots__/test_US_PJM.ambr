# serializer version: 1
# name: test_fetch_wind_solar_forecasts
  list([
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 0, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 0.0,
        'wind': 5674.163,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 1, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 0.0,
        'wind': 6406.127,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 2, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 0.0,
        'wind': 6771.246,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 3, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 0.0,
        'wind': 6846.105,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 4, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 0.0,
        'wind': 6889.084,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 5, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 0.0,
        'wind': 7046.181,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 6, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 0.0,
        'wind': 7141.57,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 7, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 0.0,
        'wind': 7272.236,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 8, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 0.0,
        'wind': 7393.612,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 9, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 0.0,
        'wind': 7533.526,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 10, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 0.0,
        'wind': 7749.602,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 11, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 15.852,
        'wind': 7933.133,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 12, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 1275.265,
        'wind': 7865.394,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 13, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 5048.272,
        'wind': 7547.31,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 14, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 7336.233,
        'wind': 7193.234,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 15, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 7802.926,
        'wind': 6876.753,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 16, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 7680.094,
        'wind': 6583.043,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 17, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 7496.111,
        'wind': 6259.908,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 18, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 7269.733,
        'wind': 5886.133,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 19, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 6945.879,
        'wind': 5548.17,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 20, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 6374.432,
        'wind': 5264.66,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 21, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 3759.787,
        'wind': 4939.546,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 22, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 637.059,
        'wind': 4831.753,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 24, 23, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'solar': 3.684,
        'wind': 5007.297,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.forecasted: 'forecasted'>,
      'storage': dict({
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
  ])
# ---
# name: test_production
  list([
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 9, 0, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'coal': 16908.0,
        'gas': 47280.0,
        'hydro': 2971.0,
        'nuclear': 32919.0,
        'oil': 871.0,
        'solar': 28.0,
        'unknown': 1838.0,
        'wind': 2782.0,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': -8.0,
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 9, 1, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'coal': 17147.0,
        'gas': 46725.0,
        'hydro': 2217.0,
        'nuclear': 32940.0,
        'oil': 837.0,
        'solar': 28.0,
        'unknown': 1641.0,
        'wind': 3202.0,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': -3.0,
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 9, 2, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'coal': 17055.0,
        'gas': 45683.0,
        'hydro': 1464.0,
        'nuclear': 32965.0,
        'oil': 566.0,
        'solar': 28.0,
        'unknown': 1551.0,
        'wind': 3504.0,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': -1.0,
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 9, 3, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'coal': 16889.0,
        'gas': 43567.0,
        'hydro': 980.0,
        'nuclear': 32987.0,
        'oil': 206.0,
        'solar': 28.0,
        'unknown': 1540.0,
        'wind': 4030.0,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': -0.0,
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 9, 4, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'coal': 16373.0,
        'gas': 42175.0,
        'hydro': 638.0,
        'nuclear': 32994.0,
        'oil': 209.0,
        'solar': 28.0,
        'unknown': 841.0,
        'wind': 4468.0,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': -0.0,
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 9, 5, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'coal': 15849.0,
        'gas': 41727.0,
        'hydro': 521.0,
        'nuclear': 33001.0,
        'oil': 206.0,
        'solar': 28.0,
        'unknown': 666.0,
        'wind': 5096.0,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': -0.0,
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 9, 6, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'coal': 15450.0,
        'gas': 40800.0,
        'hydro': 452.0,
        'nuclear': 33011.0,
        'oil': 206.0,
        'solar': 28.0,
        'unknown': 664.0,
        'wind': 5070.0,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': -0.0,
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 9, 7, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'coal': 14787.0,
        'gas': 40089.0,
        'hydro': 412.0,
        'nuclear': 33027.0,
        'oil': 206.0,
        'solar': 28.0,
        'unknown': 665.0,
        'wind': 4624.0,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': -0.0,
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 9, 8, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'coal': 14447.0,
        'gas': 39422.0,
        'hydro': 445.0,
        'nuclear': 33054.0,
        'oil': 207.0,
        'solar': 28.0,
        'unknown': 669.0,
        'wind': 4225.0,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': -0.0,
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 9, 9, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'coal': 14471.0,
        'gas': 39989.0,
        'hydro': 442.0,
        'nuclear': 33076.0,
        'oil': 209.0,
        'solar': 28.0,
        'unknown': 672.0,
        'wind': 3883.0,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': -0.0,
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 9, 10, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'coal': 14664.0,
        'gas': 40113.0,
        'hydro': 435.0,
        'nuclear': 33177.0,
        'oil': 211.0,
        'solar': 28.0,
        'unknown': 671.0,
        'wind': 3144.0,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': -0.0,
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 9, 11, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'coal': 15121.0,
        'gas': 40844.0,
        'hydro': 470.0,
        'nuclear': 33293.0,
        'oil': 212.0,
        'solar': 28.0,
        'unknown': 672.0,
        'wind': 2900.0,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': -0.0,
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 9, 12, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'coal': 15494.0,
        'gas': 41543.0,
        'hydro': 934.0,
        'nuclear': 33392.0,
        'oil': 210.0,
        'solar': 148.0,
        'unknown': 669.0,
        'wind': 2861.0,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': -0.0,
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 9, 13, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'coal': 15427.0,
        'gas': 41516.0,
        'hydro': 1015.0,
        'nuclear': 33499.0,
        'oil': 211.0,
        'solar': 2022.0,
        'unknown': 667.0,
        'wind': 2575.0,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': -0.0,
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 9, 14, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'coal': 15530.0,
        'gas': 41489.0,
        'hydro': 1080.0,
        'nuclear': 33581.0,
        'oil': 210.0,
        'solar': 4579.0,
        'unknown': 669.0,
        'wind': 2100.0,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': -0.0,
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 9, 15, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'coal': 15649.0,
        'gas': 42431.0,
        'hydro': 675.0,
        'nuclear': 33599.0,
        'oil': 211.0,
        'solar': 5452.0,
        'unknown': 668.0,
        'wind': 1630.0,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': -5.0,
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 9, 16, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'coal': 16072.0,
        'gas': 43125.0,
        'hydro': 475.0,
        'nuclear': 33613.0,
        'oil': 210.0,
        'solar': 5622.0,
        'unknown': 666.0,
        'wind': 1221.0,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': -7.0,
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 9, 17, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'coal': 16384.0,
        'gas': 42723.0,
        'hydro': 468.0,
        'nuclear': 33632.0,
        'oil': 207.0,
        'solar': 5502.0,
        'unknown': 660.0,
        'wind': 1061.0,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': -4.0,
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 9, 18, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'coal': 16411.0,
        'gas': 42256.0,
        'hydro': 496.0,
        'nuclear': 33641.0,
        'oil': 206.0,
        'solar': 5407.0,
        'unknown': 667.0,
        'wind': 1052.0,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': -0.0,
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 9, 19, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'coal': 16010.0,
        'gas': 42027.0,
        'hydro': 457.0,
        'nuclear': 33640.0,
        'oil': 205.0,
        'solar': 5231.0,
        'unknown': 724.0,
        'wind': 1187.0,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': -0.0,
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 9, 20, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'coal': 15907.0,
        'gas': 41535.0,
        'hydro': 522.0,
        'nuclear': 33629.0,
        'oil': 205.0,
        'solar': 4629.0,
        'unknown': 967.0,
        'wind': 1221.0,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': -0.0,
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2025, 2, 9, 21, 0, tzinfo=datetime.timezone.utc),
      'production': dict({
        'coal': 15772.0,
        'gas': 43252.0,
        'hydro': 1421.0,
        'nuclear': 33625.0,
        'oil': 401.0,
        'solar': 2719.0,
        'unknown': 1082.0,
        'wind': 1341.0,
      }),
      'source': 'pjm.com',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'battery': -0.0,
      }),
      'zoneKey': 'US-MIDA-PJM',
    }),
  ])
# ---
